### ISSUE TITLE
idmapped mounts - kernel 5.12

### ISSUE THREAD

**Issue by ykuksenko** [containers/podman#10374]
/kind feature

**Description**

Kernel 5.12 [merged](https://lore.kernel.org/lkml/20210213130042.828076-1-christian.brauner@ubuntu.com/T/#m8b8dc981f2e41e2b8a77d9ecb4e821c7eb977fa2) an IDMAPPED mounts feature. Is the podman project planning to make use of this?

Forgive me if this should be asked in a different location (runc/crun?). Mostly asking in case this went under the radar, I haven't seen any info about this around podman though https://github.com/containerd/containerd/pull/4734 has had some movement.

Thank you.

**Comment by rhatdan** [comment]
We are familiar with it, and are planning on using it.  We have not discussed how this will be used in RHEL yet.

**Comment by rhatdan** [comment]
@giuseppe WDYT?

**Comment by giuseppe** [comment]
AFAIK, it doesn't support yet overlay so that is not really useful at the moment.

How is containerd using the feature?  Does it work only with vfs?

**Comment by rhatdan** [comment]
We want to take an underlying image stored with the system UIDS (IE UID 0==0, 1==1 ...) and map into different containers with these UIDs remapped to match the user namespace we are running them in.  We only want to do this as readonly, if we want to write to any of these files in the image, we want overlay to copy up the files and write them with the User Namespaced UIDs.

We need to be able to do this in both rootfull (CAP_SYS_ADMIN) and rootless with User Namespace.

**Comment by giuseppe** [comment]
first of all let me rephrase "that is not really useful at the moment" -> "that is not really useful at the moment for us".

We could use it for `-v /src:/dest:O` where we currently use chown, but it won't help with mounting the container rootfs.

In our case, I think it will be easier to just use MOUNT_ATTR_IDMAP on the final overlay mount, that will be the container rootfs.

An alternative that could possibly help us is to create an idmapped bind mount for the entire storage and use it for accessing the overlay lower layers (the OCI image).  I've not investigated this alternative yet as it requires non trivial changes in the storage library so I was just waiting for overlay support since I saw it was WIP and plug it more easily in what we have now.

**Comment by brauner** [comment]
> We want to take an underlying image stored with the system UIDS (IE UID 0==0, 1==1 ...) and map into different containers with these UIDs remapped to match the user namespace we are running them in. We only want to do this as readonly, if we 

So yes, that is exactly what this is for. For the most common case where as you said we're talking about system/or init uids it's as simple as attaching the container's userns to the detached mount.

Fwiw, idmapped mounts work outside of user namespaces as well. That was an explicit use-case for systemd and others. I.e. you could remap a fileystem in the init userns with this to e.g. have real root in a privileged container write as an unprivileged uid (say `1000`) onto the disk, i.e. you could go andd create an idmapped bind-mount of your host's `/opt` at `/mnt` with a mapping of:
```
0 1000 1
1 1 1000
1001 1001 <range>`
```
and exposing this to a privileged container would mean that uid `0` aka real root creates files as uid 1000 on disk.

> want to write to any of these files in the image, we want overlay to copy up the files and write them with the User Namespaced UIDs.

Right, so this sounds like overlayfs would be mounted on top of idmapped mounts, i.e. the lower layers would be idmapped and overlayfs would be mounted from inside the container.

> 
> We need to be able to do this in both rootfull (CAP_SYS_ADMIN) and rootless with User Namespace.

Right currently MOUNT_ATTR_IDMAP requires cap_sys_admin in init_user_ns but that will change in the future. I just want to wait until later in the year before we consider this. I don't want to jump the unprivileged gun too soon.

**Comment by brauner** [comment]
> first of all let me rephrase "that is not really useful at the moment" -> "that is not really useful at the moment for us".

Thanks for that qualification. :) I appreciate that!

> 
> We could use it for `-v /src:/dest:O` where we currently use chown, but it won't help with mounting the container rootfs.

Because your rootfs mount is an overlayfs mount. Is the mount created from inside the container or is it created outside of the container?

> 
> In our case, I think it will be easier to just use MOUNT_ATTR_IDMAP on the final overlay mount, that will be the container rootfs.

Right, that's another case. That sounds like an idmapped mount of an overlayfs mount.
I have a branch that supports both, i.e. idmapped upper/lower directories and creating idmapped mounts of an overlayfs mount.

> 
> An alternative that could possibly help us is to create an idmapped bind mount for the entire storage and use it for accessing the overlay lower layers (the OCI image). I've not investigated this alternative yet as it requires non trivial changes in the storage library so I was just waiting for overlay support since I saw it was WIP and plug it more easily in what we have now.

**Comment by giuseppe** [comment]
> > We could use it for `-v /src:/dest:O` where we currently use chown, but it won't help with mounting the container rootfs.
> 
> Because your rootfs mount is an overlayfs mount. Is the mount created from inside the container or is it created outside of the container?

we create the rootfs mount outside of the container.  That is kind of forced since the OCI runtime gets just the path to the container rootfs (so it must be already mounted by Podman) and create the mount namespace later.

**Comment by brauner** [comment]
> > > We could use it for `-v /src:/dest:O` where we currently use chown, but it won't help with mounting the container rootfs.
> > 
> > 
> > Because your rootfs mount is an overlayfs mount. Is the mount created from inside the container or is it created outside of the container?
> 
> we create the rootfs mount outside of the container. That is kind of forced since the OCI runtime gets just the path to the container rootfs (so it must be already mounted by Podman) and create the mount namespace later.

Right, so the `s_user_ns`, i.e. the user namespace of the superblock is `init_user_ns` for overlayfs, right?

**Comment by giuseppe** [comment]
yes correct, at least when running as root.

For rootless we have an intermediate user+mount namespace

**Comment by giuseppe** [comment]
if you have some patches for the kernel, I can give them a try

**Comment by brauner** [comment]
> if you have some patches for the kernel, I can give them a try

I need to rebase them but I can give you something in about a week or so. I can also share a tree with you and you can poke at it as well if you want to.

**Comment by rhatdan** [comment]
@brauner did you ever make any progress on this?

**Comment by brauner** [comment]
Yes, I have indeed. My plan is to get it into shape for either next merge window or the one after that.

**Comment by giuseppe** [comment]
@brauner I've played with idmapped mounts and try to enable them only for overlay lower layers, but it seems even this configuration is not supported.




> Right, so the `s_user_ns`, i.e. the user namespace of the superblock is `init_user_ns` for overlayfs, right?

looking back at this question.  In the rootless case no, the idmapped mount should be managed inside of a user namespace.

For rootless, we create an outer user namespace with all the available IDs (as defined in `/etc/sub?id`) are mapped into.

Once in the outer user namespace, we can create inner user namespaces for each container/pod.  For these inner namespaces we'd like to reuse the container images that are pulled in the outer namespace.  The overlay mount used by the containers is created in the outer namespace.

Another question: is there any plan for supporting BTRFS or is that more complicated than XFS/ext4?

**Comment by brauner** [comment]
> @brauner I've played with idmapped mounts and try to enable them only for overlay lower layers, but it seems even this configuration is not supported.

Yes, overlayfs is currently not supported. I'm working on this. I ported another filesystem for this cycle but I've already started talking to Amir about overlayfs.

> 
> > Right, so the `s_user_ns`, i.e. the user namespace of the superblock is `init_user_ns` for overlayfs, right?
> 
> looking back at this question. In the rootless case no, the idmapped mount should be managed inside of a user namespace.
> 
> For rootless, we create an outer user namespace with all the available IDs (as defined in `/etc/sub?id`) are mapped into.



> 
> Once in the outer user namespace, we can create inner user namespaces for each container/pod. For these inner namespaces we'd like to reuse the container images that are pulled in the outer namespace. The overlay mount used by the containers is created in the outer namespace.

So essentially you're creating nested user namespaces, right? The first level is for mounting overlayfs and the second level user namespaces are for the individual containers.

Ok, so you need the ability to created idmapped mounts of an overlayfs filesystem mounted inside of a user namespace. That's a bit trickier to do but not impossible. I expected this to happen at some point anyway.

> 
> Another question: is there any plan for supporting BTRFS or is that more complicated than XFS/ext4?

That relates to why I didn't do overlayfs this cycle. :) I ported BTRFS, i.e. with 5.15 you can create idmapped mounts of BTRFS. See David Sterba's tree:

https://git.kernel.org/pub/scm/linux/kernel/git/kdave/linux.git/log/?h=for-next&qt=grep&q=brauner

and the associated test-suite specific to btrfs ioctls:

https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git/tree/src/idmapped-mounts/idmapped-mounts.c#n13423

**Comment by rhvgoyal** [comment]
@brauner So there are two pieces to it from overlayfs point of view. First one is that one should be able to use an idmapped mount as lower directory. And second piece is that one can create idmapped mounts of overlay instance. I think @giuseppe is looking for first piece to begin with. That is create idmapped mounts of ext4/xfs/btrfs/nfs and then use these mounts as lower directory of overlayfs.

Creating idmapped mounts of overlayfs might be useful in case of nested overalyfs. So if outer namespace's rootfs is on overlay and it is creating nested container then it might have to create an idmapped mount shifted as per the nested container mappings on overlayfs.

I am wondering why overlayfs does not support idmapped mount as lower dir. What are the issues.

**Comment by rhvgoyal** [comment]
And probably we will need to create idmapped mounts from inside the outer user namespace. So we can download image, create idmapped mount inside outer NS and then use that as lower dir for overlayfs mount of nested container. This does not allow image sharing between different users but atleast this will avoid chown() (or doing uid shifting using fuse-overlayfs), IIUC.

**Comment by brauner** [comment]
> @brauner So there are two pieces to it from overlayfs point of view. First one is that one should be able to use an idmapped mount as lower directory. And second piece is that one can create idmapped mounts of overlay

Yes, exactly and I did outline that a couple of times and implemented both.

instance. I think @giuseppe is looking for first piece to begin with. That is create idmapped mounts of ext4/xfs/btrfs/nfs and then use these mounts as lower directory of overlayfs.

Yes, I understand.

> 
> Creating idmapped mounts of overlayfs might be useful in case of nested overalyfs. So if outer namespace's rootfs is on overlay and it is creating nested container then it might have to create an idmapped mount shifted as per the nested container mappings on overlayfs.

Yes, that would be a second step.

> 
> I am wondering why overlayfs does not support idmapped mount as lower dir. What are the issues.

There are no technical reasons or problems why overlayfs doesn't support idmapped mounts.
First, I had idmapped mounts sit a bit in 5.13 and 5.14 so we could have some time fixing any immediate issues. So far there have been none which I attribute to the extensive test-suite that was added. 

I announced that I would prefer to do that a while ago. And now we ported `btrfs`.
While `overlayfs` will require more changes than `ext4`, `xfs`, or `btrfs` it isn't more difficult. In essence, `overlayfs` needs to be taught to take the idmapping of the underlying filesystem into account. I have been working on that but `btrfs` was higher up on the list because of `systemd` and others. I will have patches ready soon. With Plumbers and patch review and other stuff it just takes time and I do tend to spend a lot of times on tests too.

**Comment by brauner** [comment]
> And probably we will need to create idmapped mounts from inside the outer user namespace. So we can download image, create idmapped mount inside outer NS and then use that as lower dir for overlayfs mount of nested container. This does not allow image sharing between different users but atleast this will avoid chown() (or doing uid shifting using fuse-overlayfs), IIUC.

So you want to be able to create idmapped mounts inside of a user namespace, is what I understand.

**Comment by brauner** [comment]
> > Another question: is there any plan for supporting BTRFS or is that more complicated than XFS/ext4?
> 
> That relates to why I didn't do overlayfs this cycle. :) I ported BTRFS, i.e. with 5.15 you can create idmapped mounts of BTRFS. See David Sterba's tree:
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/kdave/linux.git/log/?h=for-next&qt=grep&q=brauner
> 
> and the associated test-suite specific to btrfs ioctls:
> 
> https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git/tree/src/idmapped-mounts/idmapped-mounts.c#n13423

@giuseppe https://lore.kernel.org/linux-btrfs/cover.1630327914.git.dsterba@suse.com

**Comment by rhvgoyal** [comment]
> > And probably we will need to create idmapped mounts from inside the outer user namespace. So we can download image, create idmapped mount inside outer NS and then use that as lower dir for overlayfs mount of nested container. This does not allow image sharing between different users but atleast this will avoid chown() (or doing uid shifting using fuse-overlayfs), IIUC.
> 
> So you want to be able to create idmapped mounts inside of a user namespace, is what I understand.

I think so. But @giuseppe knows it best how things are being setup in podman. My understanding is that one will download
image while running inside outer namespace and then that image will have to be idmapped mounted for nested container. And that's why there is a need to be able to do idmapped mount inside user namespace. 

@giuseppe please correct me if that's not the case.

**Comment by rhvgoyal** [comment]
> @giuseppe https://lore.kernel.org/linux-btrfs/cover.1630327914.git.dsterba@suse.com

Cool. Good to see idmapped mount for btrfs support being merged upstream.

**Comment by rhatdan** [comment]
In Rootless Mode Podman and Buildah pull the image and store it within a User Namespace, without the mapping, No Overlay.  Then the tools will optionally mount the images with and idmapping and an overlay.  Just the lower is mapped read/only, and the upper would be native on disk.

In rootless mode, the pulled image will by default run in the hosts user namespace and then mount the idmapping as a lower lower/read only as a lower with a non-mapped upper.

**Comment by giuseppe** [comment]
> @giuseppe https://lore.kernel.org/linux-btrfs/cover.1630327914.git.dsterba@suse.com

Thanks for the update! 


> In rootless mode, the pulled image will by default run in the hosts user namespace and then mount the idmapping as a lower lower/read only as a lower with a non-mapped upper.

I think this is s/rootless/root/

**Comment by rhatdan** [comment]
Yes, I will fix.

**Comment by rhatdan** [comment]
@giuseppe @baude @ykuksenko @brauner Any movement on rootless overlayfs support for idmapped mounts?

**Comment by brauner** [comment]
> @giuseppe @baude @ykuksenko @brauner Any movement on rootless overlayfs support for idmapped mounts?

Yeah, I plan to finish the big parts of it by the end of the year.

**Comment by rhatdan** [comment]
@brauner Any update we can look at?

**Comment by brauner** [comment]
> @brauner Any update we can look at?

The first part of the required vfs work for this is about to be sent this (latest next) week.

**Comment by giuseppe** [comment]
Thanks!  In the meanwhile, I've plugged support for idmapped volumes:

- https://github.com/containers/podman/pull/12298
- https://github.com/containers/crun/pull/780

**Comment by brauner** [comment]
The vfs bits are lined-up.

**Comment by rhatdan** [comment]
@giuseppe any progress on this?

**Comment by brauner** [comment]
> @giuseppe any progress on this?

VFS bits are in. So I can start on working on actually porting overlayfs this cycle.

**Comment by brauner** [comment]
Still working on this, making progress.

**Comment by giuseppe** [comment]
I am closing the issue since the kernel feature was merged yesterday and containers/storage takes advantage of it when it is present

**Comment by rhatdan** [comment]
Kernel feature for idmapped +overlay in rootful mod only, correct?

**Comment by giuseppe** [comment]
> @giuseppe could you open a bugzilla to have this backported to the RHEL9.1 kernel.

done

**Comment by jlebon** [comment]
So IIUC, with this now rootless podman container creation should become much faster from not having to `chown` everything first, right?

**Comment by rhatdan** [comment]
Containers running within a separate User Namespace should get much faster, but currently this is just for rootful containers.

$ sudo podman run --userns=auto ...

Also enables this type of functionality in CRI-O to have CRI-O launching containers in unigue User Namespaces.

idmapped is not allowed in rootless mode yet (Maybe never).

**Comment by srd424** [comment]
Apologies for a mild thread hijack, but @brauner - is there a reason why an idmapped mount of a merged overlay directory is still not allowed? I found a patch posted to linux-fsdevel back in 2020 where you'd enabled it..

**Comment by brauner** [comment]
> Apologies for a mild thread hijack, but @brauner - is there a reason why an idmapped mount of a merged overlay directory is still not allowed? I found a patch posted to linux-fsdevel back in 2020 where you'd enabled it..

There was no use-case for it. What do you want to use this for?

**Comment by srd424** [comment]
Hmm, I was thinking it would provide an easier set-up for some system container stuff until LXC/nspawn support idmapped overlays directly, but possibly I am wrong - I will try to think it through properly when I'm not full of medication. It looks like I might be able to do something with LXC pre-mount hooks instead..

**Comment by srd424** [comment]
> There was no use-case for it. What do you want to use this for?

Ah, it looks like I can simply `mount-idmapped` the layers in advance now, and then have LXC set up the overlay in the container, and it all just works. So no, I guess no real use case for it, other than completeness.

**Comment by phemmer** [comment]
https://github.com/containers/podman/issues/10374#issuecomment-844175350
> > We need to be able to do this in both rootfull (CAP_SYS_ADMIN) and rootless with User Namespace.
> 
> Right currently MOUNT_ATTR_IDMAP requires cap_sys_admin in init_user_ns but that will change in the future. I just want to wait until later in the year before we consider this. I don't want to jump the unprivileged gun too soon.

@brauner It's been a couple years now. Has there been any update or further thoughts on this?

**Comment by livingsilver94** [comment]
See this comment: https://lwn.net/Articles/896663

> Since no filesystems that support being mounted unprivileged support them

Since Linux 5.19 OverlayFS support idmapping, and an OverlayFS can be mounted unprivileged by passing the mount option `userxattr`. Thus, I think that Brauner's quote isn't valid anymore (for the sole OverlayFS at least).

I'll try to idmap an OverlayFS mount soon.

**Comment by brauner** [comment]
Nothing's really changed. While overlayfs is mountable by unprivileged users it cannot be mounted with an idmapping as taht would be really weird. Instead, you can idmap the layers and then mount overlayfs on top of the idmapped layers as an unprivileged users.

**Comment by livingsilver94** [comment]
I'm confused. See this [manpage](https://man.archlinux.org/man/mount_setattr.2.en#NOTES):

> overlayfs (ID-mapped lower and upper layers supported since Linux 5.19)

It looks like OverlayFS can be idmapped but you're saying the opposite. Please consider I'm no expert on the matter, so what I am missing?

> Instead, you can idmap the layers and then mount overlayfs on top of the idmapped layers as an unprivileged users.

That unfortunately requires CAP_SYS_ADMIN to idmap the layers, doesn't it? :(
